selection/algorithms/relaxation_algorithm.py

```diff
@@ -0,0 +1,209 @@
+from ..selection_algorithm import SelectionAlgorithm
+from ..what_if_index_creation import WhatIfIndexCreation
+from ..index import Index, index_merge, index_split
+
+import itertools
+import logging
+
+# Maximum number of columns per index, storage budget in MB,
+DEFAULT_PARAMETERS = {
+    "max_index_columns": 3,
+    "budget": 500,
+}
+
+
+class RelaxationAlgorithm(SelectionAlgorithm):
+    def __init__(self, database_connector, parameters=None):
+        if parameters is None:
+            parameters = {}
+        SelectionAlgorithm.__init__(
+            self, database_connector, parameters, DEFAULT_PARAMETERS
+        )
+        self.what_if = WhatIfIndexCreation(database_connector)
+        # convert MB to bytes
+        self.disk_constraint = self.parameters["budget"] * 1000000
+
+    def _calculate_best_indexes(self, workload):
+        logging.info("Calculating best indexes Relaxation")
+        # Obtain best indexes per query
+        _, candidates = self._exploit_virtual_indexes(workload)
+
+        # CP in Figure 5
+        cp = candidates.copy()
+        cp_size = sum(index.estimated_size for index in cp)
+        cp_cost = self.cost_evaluation.calculate_cost(workload, cp, store_size=True)
+        while cp_size > self.disk_constraint:
+            # Pick a configuration that can be relaxed
+            # TODO: Currently only one is considered
+
+            # Relax the configuration
+            best_relaxed = None
+            best_relaxed_size = None
+            lowest_relaxed_penalty = None
+
+            for transformation in ["splitting", "merging", "prefixing", "removal"]:
+                for (
+                    relaxed,
+                    relaxed_storage_savings,
+                ) in self.configurations_by_transformation(cp, transformation):
+                    relaxed_cost = self.cost_evaluation.calculate_cost(
+                        workload, relaxed, store_size=True
+                    )
+                    relaxed_cost_increase = relaxed_cost - cp_cost
+                    # TODO: Review attention
+                    # Cost savings are even better
+                    # assert (
+                    #     relaxed_cost_increase >= 0
+                    # ), f"Relaxed cost increase must be positive {transformation} {relaxed_cost_increase}"
+
+                    # TODO: Review attention
+                    if relaxed_storage_savings <= 0:
+                        continue
+                    # assert (
+                    #     relaxed_storage_savings > 0
+                    # ), f"Relaxed storage savings must be positive, {transformation}"
+                    # any storage decrease beyond the disk_constraint is not considered
```

Comment:
But if the new storage is just marginally below the `disk_constraint` we don't accept it?